README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# uncrtnty

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

{uncrtnty} provides functions and an “uncrtnty”-list class in order to
work with uncertainty in NONMEM analyses.

## Installation

You can install the development version of uncrtnty from
[github](https://github.com/FelicienLL/uncrtnty) with:

``` r
install.packages("remotes")
remotes::install_github("FelicienLL/uncrtnty")
```

## The big idea

In NONMEM analyses, the uncertainty about parameter estimation can be
described as followed. The estimation of THETA follows a multivariate
normal distribution, with a mean vector (`th_est`) and a covariance
matrix (`th_unc`) that carries the information about uncertainty. The
estimates of OMEGA/SIGMA are matrices (`om_est/si_est`), and the
uncertainty is informed by degrees of freedom (`om_unc/si_unc`). In
{uncrtnty}, these information are storred into a single “uncrtnty”-list
object.

``` r
library(uncrtnty)
u_example
#> $model
#> [1] "example001"
#> 
#> $nid
#> [1] 100
#> 
#> $nobs
#> [1] 1000
#> 
#> $th_est
#> [1] 111.00  22.00 333.00   4.44
#> 
#> $th_unc
#>      [,1] [,2]  [,3] [,4]
#> [1,] 1.11 0.12  0.13 0.14
#> [2,] 0.12 2.00  0.23 0.24
#> [3,] 0.13 0.23 33.00 0.34
#> [4,] 0.14 0.24  0.34 4.40
#> 
#> $om_est
#> $om_est[[1]]
#>      [,1] [,2]
#> [1,] 1.00 0.12
#> [2,] 0.12 2.00
#> 
#> $om_est[[2]]
#>      [,1]
#> [1,]    3
#> 
#> 
#> $om_unc
#> [1] 45 67
#> 
#> $si_est
#> $si_est[[1]]
#>      [,1] [,2]
#> [1,] 0.05    1
#> [2,] 1.00   10
#> 
#> 
#> $si_unc
#> [1] 789
#> 
#> attr(,"class")
#> [1] "uncrtnty"
```

This “uncrtnty”-list object can be created:

-   from an [xpose](https://github.com/UUPharmacometrics/xpose) database
    object: `u_from_xpdb()`.
-   from NONMEM output files (*.lst*, *.ext*, *.cov*): *available soon*.
-   from a NONMEM *.xml* output file: *available soon*.
-   from a model reported in the literature: *available soon*

This “uncrtnty”-list object can be used to generate:

-   the code implemented within the NONMEM $PRIOR NWPRI routine:
    `u_to_nwpri()`
-   the arguments to simulate with uncertainty thanks to the
    [simpar](https://github.com/metrumresearchgroup/simpar) package:
    `u_to_simpar()`

## Applications

#### Create the code for the $PRIOR routine in NONMEM

``` r
u_to_nwpri(u_example)
#> 
#> 
#> 
#> $PRIOR NWPRI
#> 
#> ;======== PRIOR BLOCKS =======
#> 
#> $THETAP ; Prior values of THETA 
#> 111  FIXED ; 1 
#> 22   FIXED ; 2 
#> 333  FIXED ; 3 
#> 4.44 FIXED ; 4 
#> 
#> $THETAPV BLOCK(4) FIXED ; Prior weight on THETA (variance-covariance matrix)
#> 1.11 ; 
#> 0.12 2    ; 
#> 0.13 0.23 33   ; 
#> 0.14 0.24 0.34 4.4  ; 
#> 
#> $OMEGAP BLOCK(2) FIXED ; Prior values of OMEGA
#> 1    ; 
#> 0.12 2    ; 
#> $OMEGAP BLOCK(1) FIXED ; Prior values of OMEGA
#> 3 ; 
#> 
#> $OMEGAPD ; Prior weight on OMEGA (degrees of freedom)
#> 45 FIXED ; 1 
#> 67 FIXED ; 2 
#> 
#> $SIGMAP BLOCK(2) FIXED ; Prior values of SIGMA
#> 0.05 ; 
#> 1    10   ; 
#> 
#> $SIGMAPD ; Prior weight on SIGMA (degrees of freedom)
#> 789 FIXED ; 1 
#> 
#> ;===== End of PRIOR BLOCKS =====
#> 
#> $COVARIANCE MATRIX = R
```

#### Create the arguments to simulate with uncrtnty in simpar

``` r
u_to_simpar(u_example, nsim = 100)
#> $nsim
#> [1] 100
#> 
#> $theta
#> [1] 111.00  22.00 333.00   4.44
#> 
#> $covar
#>      [,1] [,2]  [,3] [,4]
#> [1,] 1.11 0.12  0.13 0.14
#> [2,] 0.12 2.00  0.23 0.24
#> [3,] 0.13 0.23 33.00 0.34
#> [4,] 0.14 0.24  0.34 4.40
#> 
#> $omega
#> $omega[[1]]
#>      [,1] [,2]
#> [1,] 1.00 0.12
#> [2,] 0.12 2.00
#> 
#> $omega[[2]]
#>      [,1]
#> [1,]    3
#> 
#> 
#> $sigma
#> $sigma[[1]]
#>      [,1] [,2]
#> [1,] 0.05    1
#> [2,] 1.00   10
#> 
#> 
#> $odf
#> [1] 45 67
#> 
#> $sdf
#> [1] 789
```

## Other useful features

-   Parse information about the “Block Form” of an OMEGA/SIGMA matrix
    from:
    -   a *.lst* file with `parse_lst()`.

``` r
cat(lstfile[84:110], sep = "\n") # zoom inside a NONMEM .lst file
#> 0OMEGA HAS BLOCK FORM:
#>   1
#>   1  1
#>   0  0  2
#>   0  0  0  3
#>   0  0  0  3  3
#>   0  0  0  3  3  3
#>   0  0  0  0  0  0  4
#>   0  0  0  0  0  0  0  5
#>   0  0  0  0  0  0  0  0  6
#>   0  0  0  0  0  0  0  0  0  6
#>   0  0  0  0  0  0  0  0  0  0  7
#>   0  0  0  0  0  0  0  0  0  0  0  8
#>   0  0  0  0  0  0  0  0  0  0  0  0  9
#>   0  0  0  0  0  0  0  0  0  0  0  0  9  9
#>   0  0  0  0  0  0  0  0  0  0  0  0  0  0  9
#>   0  0  0  0  0  0  0  0  0  0  0  0  0  0  9  9
#> 0DEFAULT OMEGA BOUNDARY TEST OMITTED:    NO
#> 0SIGMA HAS BLOCK FORM:
#>   1
#>   1  1
#>   0  0  2
#>   0  0  0  3
#>   0  0  0  0  3
#>   0  0  0  0  0  4
#>   0  0  0  0  0  0  5
#> 0DEFAULT SIGMA BOUNDARY TEST OMITTED:    NO
parse_lst(lst = lstfile)$om_blockform
#>  [1] 1 1 2 3 3 3 4 5 6 6 7 8 9 9 9 9
parse_lst(lst = lstfile)$si_blockform
#> [1] 1 1 2 3 3 4 5
```

-   + the matrix itself with `infer_blockform()`.

``` r
m
#>      [,1] [,2] [,3] [,4] [,5]
#> [1,] 1.00 0.12 0.13    0    0
#> [2,] 0.12 2.00 0.23    0    0
#> [3,] 0.13 0.23 3.00    0    0
#> [4,] 0.00 0.00 0.00    4    0
#> [5,] 0.00 0.00 0.00    0    5
infer_blockform(m)
#> [1] 1 1 1 2 3
```

-   Turn a full OMEGA/SIGMA matrix into a list of OMEGA/SIGMA matrix
    blocks with `matrix_to_list()`.

``` r
matrix_to_list(m)
#> [[1]]
#>      [,1] [,2] [,3]
#> [1,] 1.00 0.12 0.13
#> [2,] 0.12 2.00 0.23
#> [3,] 0.13 0.23 3.00
#> 
#> [[2]]
#>      [,1]
#> [1,]    4
#> 
#> [[3]]
#>      [,1]
#> [1,]    5
```

-   Parse NONMEM estimate file (*.ext*), covariance matrix file (*.cov*)
    and individual estimates (*.phi*) into convenient output-list with
    `parse_ext()`, `parse_cov()` and `parse_phi()`.

-   Compute degrees of freedom of the Inverse-Wishart distribution for
    OMEGA/SIGMA matrices with `compute_df()`.

``` r
est_om <- matrix(c(0.2, 0.01, 0.01, 0.1), ncol = 2)
se_om <- matrix(c(0.02, 0.005, 0.005, 0.03), ncol = 2)
compute_df(est = est_om, se = se_om)
#> [1] 22
```


NEWS.md
# uncrtnty development version
* Fix readme and update description

# uncrtnty 0.1.2
* uncrtnty-list object are created thanks to `uncrtnty()`, that combines a constructor and a validator for more safety (see #3).
* `compute_df()` now accepts a `maxdf` to censor the maximum degree of freedom to the number of id/observations (see #2)
* Add `parse_lst()`, that can also extract the number of ID/OBS from the .lst, instead of `parse_blockform_from_lst()`
* Remove the `get_x()` functions that extracted information from xpose_data base objects
* Update README and examples

# uncrtnty 0.1.1
* The package now works on a single `uncrtnty`-list object, that is 
  - created "from": xpose data base, NONMEM outputs, etc...
  - transformed "to" : code for NONMEM PRIOR, arguments for simpar etc...
* Add `u_from_xpdb()`, `u_to_nwpri()`, `u_to_simpar()`
* Remove `xpdb_to_simpar()`
* Changes in readme
* Possibility to work on *.phi* files (`parse_phi()` and `get_phi()`)
* Add a .xml file
* Bug fix : `low_to_matrix` fill the lower matrix properly.

# uncrtnty 0.1.0.9001

* Added a `NEWS.md` file to track changes to the package.


R/to_nwpri.R
#' Turn uncrtnty-list into code for NWPRI in NONMEM
#'
#' @param u an uncrtnty-list object
#' @param cat if TRUE (the default), the character vector is passed to `cat` and printed to the console.
#' @param priors_only if TRUE, only the blocks with priors values and weight are returned.
#'
#' @return a character vector to be pasted into a NONMEM code
#' @export
#'
#' @examples
#' code <- u_to_nwpri(u_example)
#'
u_to_nwpri <- function(u, cat = TRUE, priors_only = FALSE){
  stopifnot(inherits(u, "uncrtnty"))
  ans <- paste(c(
    nwpri_value(u$th_est, blocktype = "THETAP"),
    nwpri_matrix(u$th_unc, blocktype = "THETAPV"),
    paste(sapply(u$om_est, nwpri_matrix, blocktype = "OMEGAP"), collapse = "\n"),
    nwpri_value(u$om_unc, blocktype = "OMEGAPD"),
    paste(sapply(u$si_est, nwpri_matrix, blocktype = "SIGMAP"), collapse = "\n"),
    nwpri_value(u$si_unc, blocktype = "SIGMAPD")
  ), collapse = "\n\n")

  if(!priors_only){
    prior <- "$PRIOR NWPRI"
    intro <- ";======== PRIOR BLOCKS ======="
    end <- ";===== End of PRIOR BLOCKS ====="
    cov <- "$COVARIANCE MATRIX = R"
    ans <- paste("\n", prior, intro, ans, end, cov, sep = "\n\n")
  }

  if(cat){
    cat(ans)
  }

  invisible(ans)

}

#' Set a block of matrix for the NWPRI routine in NONMEM
#'
#' @param x a symmetric matrix of numeric, typically the covariance matrix of theta or the estimates of omega/sigma.
#' @param blocktype a character, either "THETAP", "OMEGAPD" or "SIGMAPD".
#'
#' @return a single character value to be pasted into the NONMEM code
#' @export
#'
#' @examples
#' x <- matrix(c(1, 0.12, 0.13, 0.12, 22, 0.23, 0.13, 0.23, 333), nrow = 3, ncol = 3)
#' nwpri_matrix(x = x, blocktype = "THETAPV")
#' nwpri_matrix(x = x, blocktype = "OMEGAP")
nwpri_matrix <- function(x, blocktype){
  stopifnot(blocktype %in% c("THETAPV", "OMEGAP", "SIGMAP"))
  xsize <- nrow(x)
  xdescr <- nwpri_blockdescr(blocktype)
  xhead <- glue::glue("${blocktype} BLOCK({xsize}) FIXED ; {xdescr}")

  xval <- pad_matrix(x)
  xname <- colnames(x)
  if(is.null(xname)) xname <- ""
  xblock <- paste(glue::glue("{xval} ; {xname}"), collapse = "\n")

  glue::glue('{xhead}\n{xblock}')
}


#' Set a block of values for the NWPRI routine in NONMEM
#'
#' @param x a (named) vector of numeric, typically the estimates of theta or the degree of freedom associated to omega/sigma.
#' @param blocktype a character, either "THETAP", "OMEGAPD" or "SIGMAPD".
#'
#' @return a single character value to be pasted into the NONMEM code
#' @export
#'
#' @examples
#' nwpri_value(x = c(CL = .123, 456, V = 789), blocktype = "THETAP")
#' nwpri_value(x = c(123, 45, 67), blocktype = "OMEGAPD")
nwpri_value <- function(x, blocktype){
  stopifnot(blocktype %in% c("THETAP", "OMEGAPD", "SIGMAPD"))
  xdescr <- nwpri_blockdescr(blocktype)
  xhead <- glue::glue("${blocktype} ; {xdescr}")

  xval <- stringr::str_pad(x, width = max(stringr::str_length(x)), side = "right")
  xindex <- seq_len(length(x))
  xname <- names(x)
  if(is.null(xname)) xname <- ""
  xblock <- paste(glue::glue("{xval} FIXED ; {xindex} {xname}"), collapse = "\n")

  glue::glue('{xhead}\n{xblock}')
}

# pad_matrix(matrix(c(1, 0.12, 0.13, 0.12, 22, 0.23, 0.13, 0.23, 333), nrow = 3))
pad_matrix <- function(x){  #x = a symmetric matrix
  y <- character()
  pad <- max(stringr::str_length(x))
  for(i in seq_len(nrow(x))){
    y[i] <- paste(stringr::str_pad(x[i, seq_len(i)], width = pad, side = "right"), collapse = " ")
  }
  y #y = a vector of characters = rows of the lower matrix (as padded characters)
}

# nwpri_blockdescr(x = "SIGMAP")
nwpri_blockdescr <- function(x){
  blockname <- c("THETAP",
                 "THETAPV",
                 "OMEGAP",
                 "OMEGAPD",
                 "SIGMAP",
                 "SIGMAPD")

  blockdescr <- c("Prior values of THETA ",
                  "Prior weight on THETA (variance-covariance matrix)",
                  "Prior values of OMEGA",
                  "Prior weight on OMEGA (degrees of freedom)",
                  "Prior values of SIGMA",
                  "Prior weight on SIGMA (degrees of freedom)")

  blockdescr[blockname==x]
}
